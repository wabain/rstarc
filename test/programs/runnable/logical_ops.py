#!/usr/bin/env python3

binary_ops = [
    ('and', lambda x, y: x and y),
    ('or', lambda x, y: x or y),
    ('nor', lambda x, y: (not x) and (not y)),
]

binary_inputs = [
    (True, True),
    (False, True),
    (True, False),
    (False, False),
]

unary_ops = []

# Prelude
print('''\
(Test logic tables and short-circuiting)

SayTrue takes ArgId
    say ArgId
    give back true

SayFalse takes ArgId
    say ArgId
    give back false
''')

for op, expected in binary_ops:
    for (in1, in2) in binary_inputs:
        base_id = f'{op}({int(in1)},{int(in2)})'

        arg1 = 'true' if in1 else 'false'

        fn2 = 'SayTrue' if in2 else 'SayFalse'
        msg2 = f'"{base_id}:eval2"'
        arg2 = f'{fn2} taking {msg2}'

        exp = expected(in1, in2)
        true_indicator = ' ' if exp else '!'
        false_indicator = '!' if exp else ' '

        print('if', arg1, op, arg2)
        print(f'    say "{true_indicator} {base_id}:TRUE"')
        print('else')
        print(f'    say "{false_indicator} {base_id}:FALSE"')
        print()
