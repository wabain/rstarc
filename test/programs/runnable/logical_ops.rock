(NOTE: This is generated code. Apply changes to logical_ops.py and recompile)

(Test logic tables and short-circuiting)

SayTrue takes ArgId
    say ArgId
    give back true

SayFalse takes ArgId
    say ArgId
    give back false

if true and SayTrue taking "  and(1,1): eval2"
    say "  and(1,1): TRUE"
else
    say "! and(1,1): FALSE"

if false and SayTrue taking "  and(0,1): eval2"
    say "! and(0,1): TRUE"
else
    say "  and(0,1): FALSE"

if true and SayFalse taking "  and(1,0): eval2"
    say "! and(1,0): TRUE"
else
    say "  and(1,0): FALSE"

if false and SayFalse taking "  and(0,0): eval2"
    say "! and(0,0): TRUE"
else
    say "  and(0,0): FALSE"

if true or SayTrue taking "  or(1,1): eval2"
    say "  or(1,1): TRUE"
else
    say "! or(1,1): FALSE"

if false or SayTrue taking "  or(0,1): eval2"
    say "  or(0,1): TRUE"
else
    say "! or(0,1): FALSE"

if true or SayFalse taking "  or(1,0): eval2"
    say "  or(1,0): TRUE"
else
    say "! or(1,0): FALSE"

if false or SayFalse taking "  or(0,0): eval2"
    say "! or(0,0): TRUE"
else
    say "  or(0,0): FALSE"

if true nor SayTrue taking "  nor(1,1): eval2"
    say "! nor(1,1): TRUE"
else
    say "  nor(1,1): FALSE"

if false nor SayTrue taking "  nor(0,1): eval2"
    say "! nor(0,1): TRUE"
else
    say "  nor(0,1): FALSE"

if true nor SayFalse taking "  nor(1,0): eval2"
    say "! nor(1,0): TRUE"
else
    say "  nor(1,0): FALSE"

if false nor SayFalse taking "  nor(0,0): eval2"
    say "  nor(0,0): TRUE"
else
    say "! nor(0,0): FALSE"

