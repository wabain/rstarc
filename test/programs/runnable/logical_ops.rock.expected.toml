[tests]
enabled = ["run"]

[run]
returncode = "0"
stdout = """
  and(1,1): eval2
  and(1,1): TRUE
  and(0,1): FALSE
  and(1,0): eval2
  and(1,0): FALSE
  and(0,0): FALSE
  or(1,1): TRUE
  or(0,1): eval2
  or(0,1): TRUE
  or(1,0): TRUE
  or(0,0): eval2
  or(0,0): FALSE
  nor(1,1): FALSE
  nor(0,1): eval2
  nor(0,1): FALSE
  nor(1,0): FALSE
  nor(0,0): eval2
  nor(0,0): TRUE
  not(1): FALSE
  not(0): TRUE
"""
